### PR TITLE
SLE-19 implement local issue tracking to not loose creation date

### PR DESCRIPTION
**Pull request by henryju**

### CODE REVIEW

File: org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/markers/SonarMarker.java [hunk 1]
```diff
@@ -38,41 +32,67 @@
...
+      marker.setAttribute(MarkerUtils.SONAR_MARKER_CHECKSUM_ATTR, null);
+    }
+  }
 
-    marker.setAttributes(markerAttributes);
-    return this;
+  private static void updateLocationAndChecksum(final IMarker marker, final Issue issue, final IDocument iDoc, Integer startLine) throws BadLocationException, CoreException {
+    int startLineStartOffset = iDoc.getLineOffset(startLine - 1);
+    Integer issueStartLineOffset = issue.getStartLineOffset();
+    int start, end;
```
Comment (SonarTech) [inline org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/markers/SonarMarker.java]: MINOR Declare "end" on a separate line. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1659)

File: org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/jobs/AnalyzeProjectJob.java [hunk 2]
```diff
@@ -280,6 +295,151 @@ public void handle(Issue issue) {
...
+  }
+
+  private static Integer computeChecksum(IDocument iDoc, Issue issue, Integer startLine) throws BadLocationException {
+    Integer checksum;
+    if (startLine == null) {
+      checksum = null;
+    } else {
+      int startLineStartOffset = iDoc.getLineOffset(startLine - 1);
+      Integer issueStartLineOffset = issue.getStartLineOffset();
+      int start, end;
```
Comment (SonarTech) [inline org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/jobs/AnalyzeProjectJob.java]: MINOR Declare "end" on a separate line. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1659)

File: org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/jobs/AnalyzeProjectJob.java [hunk 3]
```diff
@@ -115,7 +155,64 @@ protected IStatus run(SonarRunnerFacade runner, final IProgressMonitor monitor)
...
+      iTextFileBufferManager.connect(iFile.getFullPath(), LocationKind.IFILE, new NullProgressMonitor());
+      ITextFileBuffer iTextFileBuffer = iTextFileBufferManager.getTextFileBuffer(iFile.getFullPath(), LocationKind.IFILE);
+      IDocument iDoc = iTextFileBuffer.getDocument();
+
+      issueTracking(r, previousMarkers, rawIssues, iDoc);
+
+    } finally {
+      try {
+        iTextFileBufferManager.disconnect(iFile.getFullPath(), LocationKind.IFILE, new NullProgressMonitor());
+      } catch (CoreException e) {
```
Comment (SonarTech) [inline org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/jobs/AnalyzeProjectJob.java]: CRITICAL Either log or rethrow this exception. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1166)

File: org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/tracking/Tracker.java [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+    private final Integer line;
+
+    LineAndLineHashKey(Trackable trackable) {
+      this.ruleKey = trackable.getRuleKey();
+      this.line = trackable.getLine();
+      this.lineHash = trackable.getLineHash();
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
Comment (SonarTech) [inline org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/tracking/Tracker.java]: BLOCKER Add a type test to this method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2097)

File: org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/tracking/Tracker.java [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+    private final Integer line;
+
+    LineAndMessageKey(Trackable trackable) {
+      this.ruleKey = trackable.getRuleKey();
+      this.message = trackable.getMessage();
+      this.line = trackable.getLine();
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
Comment (SonarTech) [inline org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/tracking/Tracker.java]: BLOCKER Add a type test to this method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2097)

File: org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/tracking/Tracker.java [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+        baseSearch.put(searchKey, new ArrayList<BASE>());
+      }
+      baseSearch.get(searchKey).add(base);
+    }
+
+    for (RAW raw : tracking.getUnmatchedRaws()) {
+      SearchKey rawKey = factory.create(raw);
+      Collection<BASE> bases = baseSearch.get(rawKey);
+      if (bases != null && !bases.isEmpty()) {
+        // TODO taking the first one. Could be improved if there are more than 2 issues on the same line.
```
Comment (SonarTech) [inline org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/tracking/Tracker.java]: INFO Complete the task associated to this TODO comment. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1135)

File: org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/tracking/Tracker.java [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+    private final String ruleKey;
+    private final Integer lineHash;
+
+    LineHashKey(Trackable trackable) {
+      this.ruleKey = trackable.getRuleKey();
+      this.lineHash = trackable.getLineHash();
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
Comment (SonarTech) [inline org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/tracking/Tracker.java]: BLOCKER Add a type test to this method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2097)

File: org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/tracking/Tracker.java [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+    private final Integer lineHash;
+
+    LineHashAndMessageKey(Trackable trackable) {
+      this.ruleKey = trackable.getRuleKey();
+      this.message = trackable.getMessage();
+      this.lineHash = trackable.getLineHash();
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
Comment (SonarTech) [inline org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/tracking/Tracker.java]: BLOCKER Add a type test to this method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2097)

File: org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/views/issues/CreationDateField.java [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import org.eclipse.ui.views.markers.MarkerField;
+import org.eclipse.ui.views.markers.MarkerItem;
+import org.sonarlint.eclipse.core.internal.markers.MarkerUtils;
+
+public class CreationDateField extends MarkerField {
+
+  private static final SimpleDateFormat SDF = new SimpleDateFormat();
```
Comment (SonarTech) [inline org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/views/issues/CreationDateField.java]: CRITICAL Make "SDF" an instance variable. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2885)

**Comment by SonarTech** [comment]
SonarQube analysis reported 11 issues:
* BLOCKER 4 blocker
* CRITICAL 2 critical
* MAJOR 2 major
* MINOR 2 minor
* INFO 1 info

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR Add a 'package-info.java' file to document the 'tracking' package (org.sonarsource.sonarlint-eclipse:org.sonarlint.eclipse.core:src/org/sonarlint/eclipse/core/internal/tracking) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1228)
* MAJOR [Make "write" a "static" method.](https://github.com/SonarSource/sonar-eclipse/blob/d416a226435f4c3ecdf863435f7d0955babaa105/org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/console/SonarLintConsole.java#L122) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2325)

**Comment by SonarTech** [comment]
SonarQube analysis reported 11 issues:
* BLOCKER 4 blocker
* CRITICAL 2 critical
* MAJOR 2 major
* MINOR 2 minor
* INFO 1 info

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR Add a 'package-info.java' file to document the 'tracking' package (org.sonarsource.sonarlint-eclipse:org.sonarlint.eclipse.core:src/org/sonarlint/eclipse/core/internal/tracking) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1228)
* MAJOR [Make "write" a "static" method.](https://github.com/SonarSource/sonar-eclipse/blob/d416a226435f4c3ecdf863435f7d0955babaa105/org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/console/SonarLintConsole.java#L122) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2325)

File: org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/tracking/Tracker.java (hunk 7 above)
Comment (SonarTech) [inline org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/tracking/Tracker.java]: BLOCKER Add a type test to this method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2097)

File: org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/tracking/Tracker.java (hunk 8 above)
Comment (SonarTech) [inline org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/tracking/Tracker.java]: BLOCKER Add a type test to this method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2097)

**Comment by SonarTech** [comment]
SonarQube analysis reported 8 issues:
* BLOCKER 4 blocker
* CRITICAL 1 critical
* MAJOR 2 major
* INFO 1 info

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR Add a 'package-info.java' file to document the 'tracking' package (org.sonarsource.sonarlint-eclipse:org.sonarlint.eclipse.core:src/org/sonarlint/eclipse/core/internal/tracking) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1228)
* MAJOR [Make "write" a "static" method.](https://github.com/SonarSource/sonar-eclipse/blob/ec634219d072bbc1d13cb585d36cb868730ae7b9/org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/console/SonarLintConsole.java#L122) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2325)

**Comment by SonarTech** [comment]
SonarQube analysis reported 8 issues:
* BLOCKER 4 blocker
* CRITICAL 1 critical
* MAJOR 2 major
* INFO 1 info

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR Add a 'package-info.java' file to document the 'tracking' package (org.sonarsource.sonarlint-eclipse:org.sonarlint.eclipse.core:src/org/sonarlint/eclipse/core/internal/tracking) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1228)
* MAJOR [Make "write" a "static" method.](https://github.com/SonarSource/sonar-eclipse/blob/ec634219d072bbc1d13cb585d36cb868730ae7b9/org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/console/SonarLintConsole.java#L122) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2325)

File: org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/tracking/Tracker.java (hunk 7 above)
Comment (SonarTech) [inline org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/tracking/Tracker.java]: BLOCKER Add a type test to this method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2097)

File: org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/tracking/Tracker.java (hunk 8 above)
Comment (SonarTech) [inline org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/tracking/Tracker.java]: BLOCKER Add a type test to this method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2097)

**Comment by SonarTech** [comment]
SonarQube analysis reported 8 issues:
* BLOCKER 4 blocker
* CRITICAL 1 critical
* MAJOR 2 major
* INFO 1 info

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR Add a 'package-info.java' file to document the 'tracking' package (org.sonarsource.sonarlint-eclipse:org.sonarlint.eclipse.core:src/org/sonarlint/eclipse/core/internal/tracking) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1228)
* MAJOR [Make "write" a "static" method.](https://github.com/SonarSource/sonar-eclipse/blob/06f709dc7fc3f81743758ff84a804310cfea3111/org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/console/SonarLintConsole.java#L122) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2325)

**Comment by dbmeneses** [comment]
I'm not familiar with sonarlint for eclipse, but LGTM after a quick review

**Pull request merged**